In [1]:
import pandas as pd
from dbfread import DBF
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
pd.set_option('display.max_rows', 500)

C:\Users\PC\AppData\Local\Temp\ipykernel_19116\1660992251.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
def readdbf(dbfpath):
    table = DBF(dbfpath, load=True)
    df = pd.DataFrame(iter(table))
    return df

In [4]:
df1 = readdbf('../data/Annual Area-based Poverty Map/PODES/2020/podes2020_b1b7.dbf')

df2 = readdbf('../data/Annual Area-based Poverty Map/PODES/2020/podes2020_b8b13.dbf')

df = pd.merge(df1, df2)



In [31]:
df["Primary_col"] = df['R101'] + df["R102"]
df.head()

R101 R101N R102     R102N R103           R103N R104          R104N  R107A  R107B  R107C  R301  R302A  R302B  R303  R501A1  R501A2  R501B  R502A  R502B  R503A1  R503A2  R503A3  R503A4  R503A5  R503A6  R503B  R504A1  R504A2  R504A3  R504A4  R504A5  R504B  R504C  R505  R506  R507A  R507B  R508  R509  R510  R511  R512A  R512B1  R512B2  R512B3  R601AK2  R601AK3  R601AK4  R601AK5  R601BK2  R601BK3  R601BK4  R601BK5  R601CK2  R601CK3  R601CK4  R601CK5  R601DK2  R601DK3  R601DK4  R601DK5  R601EK2  R601EK3  R601EK4  R601EK5  R601FK2  R601FK3  R601FK4  R601FK5  R601GK2  R601GK3  R601GK4  R601GK5  R601HK2  R601HK3  R601HK4  R601HK5  R601IK2  R601IK3  R601IK4  R601IK5  R601JK2  R601JK3  R601JK4  R601JK5  R601KK2  R601KK3  R601KK4  R601KK5  R602  R603AK2  R603AK3  R603AK4  R603BK2  R603BK3  R603BK4  R603CK2  R603CK3  R603CK4  R603DK2  R603DK3  R603DK4  R603EK2  R603EK3  R603EK4  R603FK2  R603FK3  R603FK4  R603GK2  R603GK3  R603GK4  R603HK2  R603HK3  R603HK4  R603IK2  R603IK3  R603IK4  R603JK2  \
0   11  ACEH   01  SIMEULUE  010  TEUPAH SELATAN  001        LATIUNG      1      3      5     1      1     10     1      68       0     10      2      1       2       4       6       7       1       4      4       2       3       6       8       2      2      3     2     3      1      3     2     3     1     0      2       0       0       0        0        0      7.0        2        0        0      1.0        1        0        0     28.0        3        1        0      0.0        0        0        0     10.0        2        0        0      1.0        1        0        0     11.0        2        0        0      2.0        2        0        0     11.0        2        0        0     11.0        2        0        0     41.0        3     2        0     36.0        3        0     99.8        4        0      9.0        2        0     36.0        3        0      9.0        2        0     36.0        3        0     36.0        3        0     99.8        4        0     11.0        2        1   
1   11  ACEH   01  SIMEULUE  010  TEUPAH SELATAN  002  LABUHAN BAJAU      1      3      5     1      1      8     1     141       0      0      2      1       2       4       6       7       1       4      4       2       3       6       8       2      2      3     2     3      1      3     2     4     1     0      2       0       0       0        0        0     41.0        3        1        1      0.0        0        0        0     19.0        3        1        0      0.0        0        0        0      2.0        1        0        0      2.0        1        0        0     17.0        3        0        0      2.0        1        0        0     17.0        3        0        0     17.0        3        0        0     41.0        3     2        0     41.0        3        0     99.8        4        0      2.0        1        0     41.0        3        0      9.0        2        0     41.0        3        0      1.0        1        0     99.8        4        0     17.0        3        1   
2   11  ACEH   01  SIMEULUE  010  TEUPAH SELATAN  003   SUAK LAMATAN      1      3      5     1      1      6     1     141       0      1      3      0       2       4       5       7       1       4      4       2       3       6       8       2      2      3     2     3      1      3     2     4     1     0      2       0       0       0        0        0     28.0        2        0        1      0.0        0        0        0     99.8        3        1        0      0.0        0        0        0      4.0        1        0        0      1.0        1        0        0      0.5        1        1        0      0.0        0        0        0      0.5        1        0        0      0.5        1        0        0     28.0        2     2        0     28.0        2        0     99.8        3        0     28.0        2        0     23.0        2        1      0.0        0        0     99.8        3        0     28.0        2        0     99.8        3        0      0.5        1        0   
3 

In [33]:
df.iloc[:,-15:].head()

R1201C  R1202  R1203  R1204  R1301AK2  R1301AK3  R1301AK4  R1301AK5  R1301BK2  R1301BK3  R1301BK4  R1301BK5  R1302A  R1302B Primary_col
0       1      1      0      1         1        52         1         5         1        40         1         5       4       3        1101
1       1      1      0      1         1        45         1         5         1        44         1         5       4       3        1101
2       1      1      0      0         1        47         1         5         1        46         1         7       3       3        1101
3       1      1      0      1         2         0         0         0         1        36         2         5       4       3        1101
4       1      1      0      1         1        49         1         5         1        57         1         5       4       3        1101

## Aggregating the data

In [28]:
agg_df = pd.read_csv("../working folder/aggregation_method_1.csv")
agg_df.head()

Kode Variabel                                              Label                                      label_english aggregate by  Unnamed: 4
0         R101N                                      NAMA PROVINSI                                      Province name           NaN        NaN
1         R102N                                NAMA KABUPATEN/KOTA                                  Regency/city name           NaN        NaN
2         R103N                                     NAMA KECAMATAN                                 Sub -district name           NaN        NaN
3         R104N                                NAMA DESA/KELURAHAN                             Village/kelurahan name           NaN        NaN
4         R107A  STATUS DEFINITIF DESA DAN OPERASIONAL DESA/KEL...  Village Definitive Status and Village/Keluraha...          mode        NaN

In [14]:
agg_df['aggregate by '].unique()

array([nan, 'mode', 'sum', 'proportion', 'mean'], dtype=object)

In [42]:
dictionary_zip = dict(zip(agg_df['Kode Variabel'], agg_df['aggregate by '].replace('nan', None)))

In [43]:
# Filter out the NaN values which represent columns not to be aggregated
agg_dict = {k: v for k, v in dictionary_zip.items() if pd.notna(v)}

def mode_agg(series):
    modes = series.mode()
    return modes[0] if not modes.empty else None

def proportion_agg(series):
    value_counts = series.value_counts(normalize=True)
    return value_counts.iloc[0] if not value_counts.empty else np.nan

In [44]:
agg_funcs = {
    'mode': mode_agg,
    'sum': 'sum',  # 'sum' can be used directly
    'proportion': proportion_agg,
    'mean' : "mean"
}
agg_dict = {k: agg_funcs[v] for k, v in agg_dict.items()}


aggregated_df = df.groupby('Primary_col').agg(agg_dict)

In [53]:
aggregated_df.reset_index(inplace=True)

In [55]:
aggregated_df.head()

Primary_col  R107A  R107B  R107C  R301  R302A  R302B  R303  R501A1  R501A2  R501B  R502A  R502B  R503A1  R503A2  R503A3  R503A4  R503A5  R503A6  R503B  R504A1  R504A2  R504A3  R504A4  R504A5  R504B  R504C      R505  R506  R507A  R507B  R508  R509      R510  R511     R512A  R512B1  R512B2  R512B3  R601AK2  R601AK3   R601AK4  R601AK5  R601BK2  R601BK3   R601BK4  R601BK5  R601CK2  R601CK3    R601CK4  R601CK5  R601DK2  R601DK3   R601DK4  R601DK5  R601EK2  R601EK3    R601EK4  R601EK5  R601FK2  R601FK3   R601FK4  R601FK5  R601GK2  R601GK3    R601GK4  R601GK5  R601HK2  R601HK3   R601HK4  R601HK5  R601IK2  R601IK3    R601IK4  R601IK5  R601JK2  R601JK3    R601JK4  R601JK5  R601KK2  R601KK3    R601KK4  R601KK5      R602  R603AK2    R603AK3  R603AK4  R603BK2    R603BK3  R603BK4  R603CK2    R603CK3  R603CK4  R603DK2    R603DK3  R603DK4  R603EK2   R603EK3  R603EK4  R603FK2    R603FK3  R603FK4  R603GK2    R603GK3  R603GK4  R603HK2    R603HK3  R603HK4  R603IK2    R603IK3  R603IK4  R603JK2  \
0        1101      1      3      5     1      1   2544     1   23319      38    525      2      1       2       4       6       7       1       4      4       2       3       6       8       2      2      3  0.992754     3      1      1     2     4  0.818841    27  1.000000       0       0       0       21       62  5.563768        0       33       58  3.581884        0        0        2  41.005072        3      114        2  1.080435        0        6        6  18.505072        2       44        3  1.772464        2        2        9  10.007971        2       25        1  4.047826        2        1        5  25.905072        2        9        0  14.875362        2        0        1  49.129710        3  0.985507        1  43.824638        3        0  78.976087        3       14   7.007971        2        5  39.410145        2       61  1.956522        0        1  50.334783        3        8  40.763696        3        0  74.186739        3       16  25.156377        2       74   
1        1102      1      3      5     1      1    818     2   30622     149    308      2      1       2       3       6       8       1       4      2       2       3       6       8       2      2      3  0.958333     3      1      1     2     4  0.766667    17  0.975000      12     143     136       28      118  1.480833        0       21      118  0.620000        0        0       14  16.987917        2      106        5  0.460000        0        2       28   9.729750        2       33        8  2.368250        2        1        8  12.328750        2       11        2  5.463167        2        1        7  13.057083        2        6        3  10.372250        2        0        4  18.236667        2  0.841667        1  21.055583        2        0  63.517500        2        7  10.394000        2        7   7.889083        2       34  5.081500        2        6  16.629583        2       23  11.830417        2        4  41.091250        3       37   7.905583        2       53   
2        1103      1      3      5     1      1   3516     2   63310     134    663      1      1       2       3       5       8       1       4      2       2       3       6       8       2      2      3  0.996154     3      1      1     2     4  0.784615     6  1.000000       0       0       0       63      232  0.013462        0       52       97  0.985769        0        3        6  24.820769        2      202        6  0.332192        0       28        8   9.624231        2       52        7  1.544231        2        6       25   4.874615        2       28        6  2.995962        2        4       12   9.885385        2       10        1   6.863077        2        2        2  40.611538        2  0.861538        1  34.729231        2        0  84.938462        4       21   3.888846        2        6  16.960577        2       52  6.002115        2        7  23.168077        2       22  10.885000        2       15  15.816154        2       52   4.816923        2       48   
3        1104      1      

In [54]:
aggregated_df.describe()

R107A  R107B  R107C        R301       R302A        R302B        R303        R501A1        R501A2          R501B       R502A       R502B      R503A1      R503A2      R503A3      R503A4      R503A5      R503A6       R503B      R504A1      R504A2      R504A3      R504A4      R504A5       R504B       R504C        R505        R506       R507A       R507B        R508        R509        R510        R511       R512A      R512B1        R512B2        R512B3     R601AK2      R601AK3     R601AK4     R601AK5     R601BK2      R601BK3     R601BK4     R601BK5     R601CK2     R601CK3     R601CK4     R601CK5      R601DK2     R601DK3     R601DK4     R601DK5     R601EK2     R601EK3     R601EK4     R601EK5     R601FK2     R601FK3     R601FK4     R601FK5     R601GK2     R601GK3     R601GK4     R601GK5     R601HK2     R601HK3     R601HK4     R601HK5     R601IK2     R601IK3     R601IK4     R601IK5     R601JK2     R601JK3     R601JK4     R601JK5     R601KK2     R601KK3     R601KK4     R601KK5  \
count  514.0  514.0  514.0  514.000000  514.000000   514.000000  514.000000  5.140000e+02    514.000000     514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000    514.000000    514.000000  514.000000   514.000000  514.000000  514.000000  514.000000   514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000   514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000  514.000000   
mean     1.0    3.0    5.0    1.225681    1.009728  1052.239300    1.365759  1.545850e+05   3269.490272    2528.490272    1.688716    0.850195    1.988327    3.194553    5.359922    7.643969    1.089494    3.986381    2.618677    1.725681    3.087549    5.848249    7.990272    1.968872    1.896887    2.665370    0.908577    2.836576    1.089494    1.311284    4.894942    3.659533    0.822230   48.453307    0.932915   21.616732   1058.215953   1421.947471   36.468872   246.657588    3.930777    0.249027   13.000000   189.538911    4.818665    0.511673    0.729572   59.867704   21.424124    1.815175   257.807393   35.097276    0.599172    0.077821    4.081712   52.303502   18.161984    1.770428   46.795720   33.165370    3.246837    0.933852    3.496109   34.383268   16.318037    1.875486   13.904669   14.758755    8.122519    1.811284    1.768482   16.643969   21.654080    2.182879    7.328794   20.702335   13.396166    1.906615    1.293774    7.700389   32.607284    2.291829   
std      0.0    0.0    0.0    0.551130    0.098243   881.358001    0.482111  2.015165e+05   6777.363068    9456.157334    0.772574    0.457694    0.107515    0.396241    0.480445    0.479291    0.285734    0.116015    1.184089    0.446605    0.282913    0.359129    0.098243    0.173834    0.694640    0.747241    0.116609    0.548385    0.491414    0.762415    2.253145    1.458602    0.131489   87.672852    0.110831   42.513878   3162.453267   4563.486269   44.691317   287.437267   10.991681    0.849406   12.638164   207.075815   10.329049    1.047070    1.686670  107.066420   26.174643    1.208677   230.913365   51.324772    2.674617    0.430076    6.223669   94.042011   24.670330    1.162197   28.687927   56.287083    5.756137    1.050156    3.648945   54.129681   23.991320    1.020292    8.654549   23.963995    9.820742    0.763053    1.648113   26.103552   26.622153    0.795711    4.533057   34.779715   15.898755    0.864637    2.351492   13.392694   27.681219    0.876625   
min      1.0    3.0    5.0    1.000

In [56]:
aggregated_df.shape

(514, 188)

In [88]:
crosswalk = pd.read_excel("../working folder/Crosswalk_DAU@3.xlsx")
crosswalk.head()

area_eng           area_bahind   area_cl_bahind  _ID        _CX       _CY        shapeName  name_simple_bps           fao_adm_name                 wb_combined_names  bps_code_all  prov_all  kab_all              name_all  bps_code                  name  prov   kab
0       Regency. West Aceh       Kab. Aceh Barat       Aceh Barat  1.0  96.185453  4.456643       Aceh Barat       Aceh Barat       Aceh Barat, Kab.       Indonesia, Aceh Barat, Kab.        1107.0      11.0      7.0       Kab. Aceh Barat    1107.0       Kab. Aceh Barat  11.0   7.0
1  Regency. Southwest Aceh  Kab. Aceh Barat Daya  Aceh Barat Daya  2.0  96.878320  3.824899  Aceh Barat Daya  Aceh Barat Daya  Aceh Barat Daya, Kab.  Indonesia, Aceh Barat Daya, Kab.        1112.0      11.0     12.0  Kab. Aceh Barat Daya    1112.0  Kab. Aceh Barat Daya  11.0  12.0
2      Regency. Aceh Besar       Kab. Aceh Besar       Aceh Besar  3.0  95.515608  5.379938       Aceh Besar       Aceh Besar       Aceh Besar, Kab.       Indonesia, Aceh Besar, Kab.        1108.0      11.0      8.0       Kab. Aceh Besar    1108.0       Kab. Aceh Besar  11.0   8.0
3       Regency. Aceh Jaya        Kab. Aceh Jaya        Aceh Jaya  4.0  95.678706  4.832184        Aceh Jaya        Aceh Jaya        Aceh Jaya, Kab.        Indonesia, Aceh Jaya, Kab.        1116.0      11.0     16.0        Kab. Aceh Jaya    1116.0        Kab. Aceh Jaya  11.0  16.0
4      Regency. South Aceh     Kab. Aceh Selatan     Aceh Selatan  5.0  97.435351  3.162864     Aceh Selatan     Aceh Selatan     Aceh Selatan, Kab.     Indonesia, Aceh Selatan, Kab.        1103.0      11.0      3.0     Kab. Aceh Selatan    1103.0     Kab. Aceh Selatan  11.0   3.0

## 34 BPS codes have null values 

and cannot be joined with the podus aggregated dataframe

In [89]:
crosswalk['bps_code'].isnull().sum()

34

In [95]:
crosswalk.dropna(subset=['bps_code'], inplace=True)
crosswalk['bps_code'] = crosswalk['bps_code'].astype(int).astype(str)
crosswalk.head()


area_eng           area_bahind   area_cl_bahind  _ID        _CX       _CY        shapeName  name_simple_bps           fao_adm_name                 wb_combined_names  bps_code_all  prov_all  kab_all              name_all bps_code                  name  prov   kab
0       Regency. West Aceh       Kab. Aceh Barat       Aceh Barat  1.0  96.185453  4.456643       Aceh Barat       Aceh Barat       Aceh Barat, Kab.       Indonesia, Aceh Barat, Kab.        1107.0      11.0      7.0       Kab. Aceh Barat     1107       Kab. Aceh Barat  11.0   7.0
1  Regency. Southwest Aceh  Kab. Aceh Barat Daya  Aceh Barat Daya  2.0  96.878320  3.824899  Aceh Barat Daya  Aceh Barat Daya  Aceh Barat Daya, Kab.  Indonesia, Aceh Barat Daya, Kab.        1112.0      11.0     12.0  Kab. Aceh Barat Daya     1112  Kab. Aceh Barat Daya  11.0  12.0
2      Regency. Aceh Besar       Kab. Aceh Besar       Aceh Besar  3.0  95.515608  5.379938       Aceh Besar       Aceh Besar       Aceh Besar, Kab.       Indonesia, Aceh Besar, Kab.        1108.0      11.0      8.0       Kab. Aceh Besar     1108       Kab. Aceh Besar  11.0   8.0
3       Regency. Aceh Jaya        Kab. Aceh Jaya        Aceh Jaya  4.0  95.678706  4.832184        Aceh Jaya        Aceh Jaya        Aceh Jaya, Kab.        Indonesia, Aceh Jaya, Kab.        1116.0      11.0     16.0        Kab. Aceh Jaya     1116        Kab. Aceh Jaya  11.0  16.0
4      Regency. South Aceh     Kab. Aceh Selatan     Aceh Selatan  5.0  97.435351  3.162864     Aceh Selatan     Aceh Selatan     Aceh Selatan, Kab.     Indonesia, Aceh Selatan, Kab.        1103.0      11.0      3.0     Kab. Aceh Selatan     1103     Kab. Aceh Selatan  11.0   3.0

In [97]:
## BPS code and primary col comparision

full_list = aggregated_df['Primary_col'].unique()
crosswalk_bps = crosswalk['bps_code'].unique()
missing_numbers = sorted(set(full_list) - set(crosswalk_bps))
missing_numbers

['1109', '3101', '3171', '3172', '3173', '3174', '3175', '7301']

### In addition to 34 rows where there no bps code, tthere arre 8 bps codes missing that are present in podus data

In [112]:
podus_crosswalk = aggregated_df.merge(crosswalk, left_on='Primary_col', right_on='bps_code', how='left')
podus_crosswalk.head()

Primary_col  R107A  R107B  R107C  R301  R302A  R302B  R303  R501A1  R501A2  R501B  R502A  R502B  R503A1  R503A2  R503A3  R503A4  R503A5  R503A6  R503B  R504A1  R504A2  R504A3  R504A4  R504A5  R504B  R504C      R505  R506  R507A  R507B  R508  R509      R510  R511     R512A  R512B1  R512B2  R512B3  R601AK2  R601AK3   R601AK4  R601AK5  R601BK2  R601BK3   R601BK4  R601BK5  R601CK2  R601CK3    R601CK4  R601CK5  R601DK2  R601DK3   R601DK4  R601DK5  R601EK2  R601EK3    R601EK4  R601EK5  R601FK2  R601FK3   R601FK4  R601FK5  R601GK2  R601GK3    R601GK4  R601GK5  R601HK2  R601HK3   R601HK4  R601HK5  R601IK2  R601IK3    R601IK4  R601IK5  R601JK2  R601JK3    R601JK4  R601JK5  R601KK2  R601KK3    R601KK4  R601KK5      R602  R603AK2    R603AK3  R603AK4  R603BK2    R603BK3  R603BK4  R603CK2    R603CK3  R603CK4  R603DK2    R603DK3  R603DK4  R603EK2   R603EK3  R603EK4  R603FK2    R603FK3  R603FK4  R603GK2    R603GK3  R603GK4  R603HK2    R603HK3  R603HK4  R603IK2    R603IK3  R603IK4  R603JK2  \
0        1101      1      3      5     1      1   2544     1   23319      38    525      2      1       2       4       6       7       1       4      4       2       3       6       8       2      2      3  0.992754     3      1      1     2     4  0.818841    27  1.000000       0       0       0       21       62  5.563768        0       33       58  3.581884        0        0        2  41.005072        3      114        2  1.080435        0        6        6  18.505072        2       44        3  1.772464        2        2        9  10.007971        2       25        1  4.047826        2        1        5  25.905072        2        9        0  14.875362        2        0        1  49.129710        3  0.985507        1  43.824638        3        0  78.976087        3       14   7.007971        2        5  39.410145        2       61  1.956522        0        1  50.334783        3        8  40.763696        3        0  74.186739        3       16  25.156377        2       74   
1        1102      1      3      5     1      1    818     2   30622     149    308      2      1       2       3       6       8       1       4      2       2       3       6       8       2      2      3  0.958333     3      1      1     2     4  0.766667    17  0.975000      12     143     136       28      118  1.480833        0       21      118  0.620000        0        0       14  16.987917        2      106        5  0.460000        0        2       28   9.729750        2       33        8  2.368250        2        1        8  12.328750        2       11        2  5.463167        2        1        7  13.057083        2        6        3  10.372250        2        0        4  18.236667        2  0.841667        1  21.055583        2        0  63.517500        2        7  10.394000        2        7   7.889083        2       34  5.081500        2        6  16.629583        2       23  11.830417        2        4  41.091250        3       37   7.905583        2       53   
2        1103      1      3      5     1      1   3516     2   63310     134    663      1      1       2       3       5       8       1       4      2       2       3       6       8       2      2      3  0.996154     3      1      1     2     4  0.784615     6  1.000000       0       0       0       63      232  0.013462        0       52       97  0.985769        0        3        6  24.820769        2      202        6  0.332192        0       28        8   9.624231        2       52        7  1.544231        2        6       25   4.874615        2       28        6  2.995962        2        4       12   9.885385        2       10        1   6.863077        2        2        2  40.611538        2  0.861538        1  34.729231        2        0  84.938462        4       21   3.888846        2        6  16.960577        2       52  6.002115        2        7  23.168077        2       22  10.885000        2       15  15.816154        2       52   4.816923        2       48   
3        1104      1      

In [113]:
podus_crosswalk.describe(include='all') 

Primary_col  R107A  R107B  R107C        R301       R302A        R302B        R303        R501A1        R501A2          R501B       R502A       R502B      R503A1      R503A2      R503A3      R503A4      R503A5      R503A6       R503B      R504A1      R504A2      R504A3      R504A4      R504A5       R504B       R504C        R505        R506       R507A       R507B        R508        R509        R510        R511       R512A      R512B1        R512B2        R512B3     R601AK2      R601AK3     R601AK4     R601AK5     R601BK2      R601BK3     R601BK4     R601BK5     R601CK2     R601CK3     R601CK4     R601CK5      R601DK2     R601DK3     R601DK4     R601DK5     R601EK2     R601EK3     R601EK4     R601EK5     R601FK2     R601FK3     R601FK4     R601FK5     R601GK2     R601GK3     R601GK4     R601GK5     R601HK2     R601HK3     R601HK4     R601HK5     R601IK2     R601IK3     R601IK4     R601IK5     R601JK2     R601JK3     R601JK4     R601JK5     R601KK2     R601KK3     R601KK4  \
count          516  516.0  516.0  516.0  516.000000  516.000000   516.000000  516.000000  5.160000e+02    516.000000     516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000    516.000000    516.000000  516.000000   516.000000  516.000000  516.000000  516.000000   516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000   516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000   
unique         514    NaN    NaN    NaN         NaN         NaN          NaN         NaN           NaN           NaN            NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN           NaN           NaN         NaN          NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN   
top           6104    NaN    NaN    NaN         NaN         NaN          NaN         NaN           NaN           NaN            NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN           NaN           NaN         NaN          NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN   
freq             2    NaN    Na

### In the end only 508 observations from crosswalk could be joined with podus data, instead of 516

In [120]:
geo = pd.read_csv("../data/geospatial data/65e5e020f2269b448315b4b2_results.csv")
geo['asdf_id'] = geo['asdf_id']+1
geo['asdf_id'] = geo['asdf_id'].astype(str)
geo.head()

asdf_id  worldbank_geocodedresearchrelease_level1_v1_4_2.fa137b9.sum  globalwindatlas_powerdensity.none.mean  globalwindatlas_powerdensity.none.max  globalwindatlas_powerdensity.none.min Level  gqid  id shapeGroup                     shapeID        shapeName shapeType
0       1                                       5.559942e+07                                         14.294998                             454.602234                               0.088190  ADM2     0   0        IDN  IDN-ADM2-27142891B94608663       Aceh Barat      ADM2
1       2                                       3.266010e+07                                         13.189348                             295.704468                               0.001019  ADM2     1   1        IDN  IDN-ADM2-27142891B24353237  Aceh Barat Daya      ADM2
2       3                                       7.205353e+07                                        109.074803                            5247.738770                               0.436401  ADM2     2   2        IDN    IDN-ADM2-27142891B117820       Aceh Besar      ADM2
3       4                                       6.693449e+07                                         43.830171                            3782.134521                               0.092405  ADM2     3   3        IDN  IDN-ADM2-27142891B74934579        Aceh Jaya      ADM2
4       5                                       8.047280e+07                                         19.909913                             400.697266                               0.007878  ADM2     4   4        IDN  IDN-ADM2-27142891B88611210     Aceh Selatan      ADM2

In [121]:
podus_crosswalk['_ID'].fillna("999", inplace=True)
podus_crosswalk['_ID'] = podus_crosswalk['_ID'].astype(int).astype(str)



C:\Users\PC\AppData\Local\Temp\ipykernel_19116\476520756.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  podus_crosswalk['_ID'].fillna("999", inplace=True)


In [124]:
## podus_crosswalk _ID and geo asdf_id comparision

full_list = podus_crosswalk['_ID'].unique()
asdf_id = geo['asdf_id'].unique()
missing_numbers = sorted(set(asdf_id) - set(full_list))
missing_numbers

['124', '159', '203', '204', '205', '206', '207', '510', '511', '91', '92']

### 11 asdf_id's are missing from the podus_crosswalk aggregated data

In [123]:
final_df = podus_crosswalk.merge(geo, left_on='_ID', right_on='asdf_id', how='left')
print(final_df.shape)
final_df.head()

(516, 218)


Primary_col  R107A  R107B  R107C  R301  R302A  R302B  R303  R501A1  R501A2  R501B  R502A  R502B  R503A1  R503A2  R503A3  R503A4  R503A5  R503A6  R503B  R504A1  R504A2  R504A3  R504A4  R504A5  R504B  R504C      R505  R506  R507A  R507B  R508  R509      R510  R511     R512A  R512B1  R512B2  R512B3  R601AK2  R601AK3   R601AK4  R601AK5  R601BK2  R601BK3   R601BK4  R601BK5  R601CK2  R601CK3    R601CK4  R601CK5  R601DK2  R601DK3   R601DK4  R601DK5  R601EK2  R601EK3    R601EK4  R601EK5  R601FK2  R601FK3   R601FK4  R601FK5  R601GK2  R601GK3    R601GK4  R601GK5  R601HK2  R601HK3   R601HK4  R601HK5  R601IK2  R601IK3    R601IK4  R601IK5  R601JK2  R601JK3    R601JK4  R601JK5  R601KK2  R601KK3    R601KK4  R601KK5      R602  R603AK2    R603AK3  R603AK4  R603BK2    R603BK3  R603BK4  R603CK2    R603CK3  R603CK4  R603DK2    R603DK3  R603DK4  R603EK2   R603EK3  R603EK4  R603FK2    R603FK3  R603FK4  R603GK2    R603GK3  R603GK4  R603HK2    R603HK3  R603HK4  R603IK2    R603IK3  R603IK4  R603JK2  \
0        1101      1      3      5     1      1   2544     1   23319      38    525      2      1       2       4       6       7       1       4      4       2       3       6       8       2      2      3  0.992754     3      1      1     2     4  0.818841    27  1.000000       0       0       0       21       62  5.563768        0       33       58  3.581884        0        0        2  41.005072        3      114        2  1.080435        0        6        6  18.505072        2       44        3  1.772464        2        2        9  10.007971        2       25        1  4.047826        2        1        5  25.905072        2        9        0  14.875362        2        0        1  49.129710        3  0.985507        1  43.824638        3        0  78.976087        3       14   7.007971        2        5  39.410145        2       61  1.956522        0        1  50.334783        3        8  40.763696        3        0  74.186739        3       16  25.156377        2       74   
1        1102      1      3      5     1      1    818     2   30622     149    308      2      1       2       3       6       8       1       4      2       2       3       6       8       2      2      3  0.958333     3      1      1     2     4  0.766667    17  0.975000      12     143     136       28      118  1.480833        0       21      118  0.620000        0        0       14  16.987917        2      106        5  0.460000        0        2       28   9.729750        2       33        8  2.368250        2        1        8  12.328750        2       11        2  5.463167        2        1        7  13.057083        2        6        3  10.372250        2        0        4  18.236667        2  0.841667        1  21.055583        2        0  63.517500        2        7  10.394000        2        7   7.889083        2       34  5.081500        2        6  16.629583        2       23  11.830417        2        4  41.091250        3       37   7.905583        2       53   
2        1103      1      3      5     1      1   3516     2   63310     134    663      1      1       2       3       5       8       1       4      2       2       3       6       8       2      2      3  0.996154     3      1      1     2     4  0.784615     6  1.000000       0       0       0       63      232  0.013462        0       52       97  0.985769        0        3        6  24.820769        2      202        6  0.332192        0       28        8   9.624231        2       52        7  1.544231        2        6       25   4.874615        2       28        6  2.995962        2        4       12   9.885385        2       10        1   6.863077        2        2        2  40.611538        2  0.861538        1  34.729231        2        0  84.938462        4       21   3.888846        2        6  16.960577        2       52  6.002115        2        7  23.168077        2       22  10.885000        2       15  15.816154        2       52   4.816923        2       48   
3        1104      1      

In [125]:
final_df.describe(include='all')    

Primary_col  R107A  R107B  R107C        R301       R302A        R302B        R303        R501A1        R501A2          R501B       R502A       R502B      R503A1      R503A2      R503A3      R503A4      R503A5      R503A6       R503B      R504A1      R504A2      R504A3      R504A4      R504A5       R504B       R504C        R505        R506       R507A       R507B        R508        R509        R510        R511       R512A      R512B1        R512B2        R512B3     R601AK2      R601AK3     R601AK4     R601AK5     R601BK2      R601BK3     R601BK4     R601BK5     R601CK2     R601CK3     R601CK4     R601CK5      R601DK2     R601DK3     R601DK4     R601DK5     R601EK2     R601EK3     R601EK4     R601EK5     R601FK2     R601FK3     R601FK4     R601FK5     R601GK2     R601GK3     R601GK4     R601GK5     R601HK2     R601HK3     R601HK4     R601HK5     R601IK2     R601IK3     R601IK4     R601IK5     R601JK2     R601JK3     R601JK4     R601JK5     R601KK2     R601KK3     R601KK4  \
count          516  516.0  516.0  516.0  516.000000  516.000000   516.000000  516.000000  5.160000e+02    516.000000     516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000    516.000000    516.000000  516.000000   516.000000  516.000000  516.000000  516.000000   516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000   516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000  516.000000   
unique         514    NaN    NaN    NaN         NaN         NaN          NaN         NaN           NaN           NaN            NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN           NaN           NaN         NaN          NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN   
top           6104    NaN    NaN    NaN         NaN         NaN          NaN         NaN           NaN           NaN            NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN           NaN           NaN         NaN          NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN          NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN   
freq             2    NaN    Na